firebase-firestore/src/main/java/com/google/firebase/firestore/model/value/ProtoValues.java

```diff
@@ -166,39 +166,46 @@ public static int compare(Value left, Value right) {
     }
   }
 
-  private static int compareMaps(Value left, Value right) {
-    Iterator<Map.Entry<String, Value>> iterator1 =
-        new TreeMap<>(left.getMapValue().getFieldsMap()).entrySet().iterator();
-    Iterator<Map.Entry<String, Value>> iterator2 =
-        new TreeMap<>(right.getMapValue().getFieldsMap()).entrySet().iterator();
-    while (iterator1.hasNext() && iterator2.hasNext()) {
-      Map.Entry<String, Value> entry1 = iterator1.next();
-      Map.Entry<String, Value> entry2 = iterator2.next();
-      int keyCompare = entry1.getKey().compareTo(entry2.getKey());
-      if (keyCompare != 0) {
-        return keyCompare;
+  private static int compareNumbers(Value left, Value right) {
+    if (left.getValueTypeCase() == Value.ValueTypeCase.DOUBLE_VALUE) {
+      double thisDouble = left.getDoubleValue();
+      if (right.getValueTypeCase() == Value.ValueTypeCase.DOUBLE_VALUE) {
+        return Util.compareDoubles(thisDouble, right.getDoubleValue());
+      } else if (right.getValueTypeCase() == Value.ValueTypeCase.INTEGER_VALUE) {
+        return Util.compareMixed(thisDouble, right.getIntegerValue());
       }
-      int valueCompare = compare(entry1.getValue(), entry2.getValue());
-      if (valueCompare != 0) {
-        return valueCompare;
+    } else if (left.getValueTypeCase() == Value.ValueTypeCase.INTEGER_VALUE) {
+      long thisLong = left.getIntegerValue();
+      if (right.getValueTypeCase() == Value.ValueTypeCase.INTEGER_VALUE) {
+        return Util.compareLongs(thisLong, right.getIntegerValue());
+      } else if (right.getValueTypeCase() == Value.ValueTypeCase.DOUBLE_VALUE) {
+        return -1 * Util.compareMixed(right.getDoubleValue(), thisLong);
       }
     }
 
-    // Only equal if both iterators are exhausted.
-    return Util.compareBooleans(iterator1.hasNext(), iterator2.hasNext());
+    throw fail("Unexpected values: %s vs %s", left, right);
   }
 
-  private static int compareArrays(Value left, Value right) {
-    int minLength =
-        Math.min(left.getArrayValue().getValuesCount(), right.getArrayValue().getValuesCount());
+  private static int compareTimestamps(Value left, Value right) {
+    if (left.getTimestampValue().getSeconds() == right.getTimestampValue().getSeconds()) {
```

Comment:
Elsewhere we compare and then if non-zero do the second one. Consider rephrasing as:

```java
int cmp = Util.compareIntegers(left.getTimestampValue().getSeconds(), right.getTimestampValue().getSeconds());
if (cmp != 0) return cmp;

return Util.compareIntegers(left.getTimestampValue().getNanos(), right.getTimestampValue().getNanos());
```

This has two benefits:
  * it makes the hierarchy of comparison linear; and
  * it avoids using subtraction for comparison, which is specious in the general case due to the possibility of overflow. I think it's the case that timestamp values are defined in such a way that subtraction can't possibly overflow, but there's no reason not to reuse the comparator we already have, in which case we don't have to think about it.

Reply:
This code is actually based on the original code. I do think the comparisons in the other functions are easier to follow, so I updated it.

As part of this, I changed the method to take TimestampValues (along with the rest). This avoids a bunch of extra linebreaks since it removes the calls to getTimestampValue() from here.